File: java/src/funcsplayer3/Splasher.java
package funcsplayer3;

import java.util.Random;
import battlecode.common.*;

public class Splasher extends Robot {
	
	static RobotController rc;
	static MapLocation nearestTower = null;
	static MapLocation prevTarget;
	static MapLocation my_target = null;
	static boolean reached_target = false;
	static boolean hasResource = false;
	static boolean hasWall = false;
	
	static boolean towerSpotted = false;
	static MapLocation myEnemyTower = null;

    Splasher(RobotController rc) throws GameActionException {
        super(rc);
        Splasher.rc = rc;
        
    }

    void play() throws GameActionException {
    	funcsplayer3.Util.init(rc);
    	funcsplayer3.Explore.init(rc);
    	funcsplayer3.BugNav.init(rc);
        funcsplayer3.Pathfinding.init(rc);
        funcsplayer3.Pathfinding.initTurn();
        getTarget();
        while (true) {
        	MapLocation closest_srp = new MapLocation(rc.getLocation().x - (rc.getLocation().x % 4) + 2, rc.getLocation().y - (rc.getLocation().y % 4) + 2);
    		if (rc.canCompleteResourcePattern(closest_srp)) {
    			rc.completeResourcePattern(closest_srp);
    		}
    		
        	if(towerSpotted){
        		prevTarget = myEnemyTower;
    			if (rc.canAttack(myEnemyTower)){
    				rc.attack(myEnemyTower);
    			}
    			else if(rc.isActionReady()){
    				funcsplayer3.Pathfinding.move(myEnemyTower, true);
    			}
    			if (rc.canAttack(myEnemyTower)){
    				rc.attack(myEnemyTower);
    			}
    			if (rc.canAttack(myEnemyTower)){
    				rc.attack(myEnemyTower);
    			}
    			if (rc.canAttack(myEnemyTower)){
    				rc.attack(myEnemyTower);
    			}
    			if (rc.senseRobotAtLocation(myEnemyTower)==null){
    				towerSpotted = false;
    			}
    			Direction dir = rc.getLocation().directionTo(myEnemyTower).opposite();
    			if (rc.canMove(dir)){
    				rc.move(dir);
    				rc.setIndicatorString("GOING BACK");
    			}
    			return;
    		}
    		
        	
        	MapInfo[] nearbyTiles = rc.senseNearbyMapInfos();
        	// boolean moveAway = false;
        	for (MapInfo tile : nearbyTiles){
        		if (rc.canCompleteResourcePattern(tile.getMapLocation())) {
        			rc.completeResourcePattern(tile.getMapLocation());
        		}
        		RobotInfo potentialTower = rc.senseRobotAtLocation(tile.getMapLocation());
    			if (tile.hasRuin() && potentialTower != null) {
    				if (potentialTower.getTeam() != rc.getTeam()) {
    					// enemy tower
    					myEnemyTower = potentialTower.location;
    					Direction dir = rc.getLocation().directionTo(myEnemyTower);
    					towerSpotted = true;
    					if(rc.canMove(dir)){
    						rc.move(dir);
    					}
    					if (rc.canAttack(myEnemyTower)){
    						rc.attack(myEnemyTower);
    					}
    					if (rc.canAttack(myEnemyTower)){
    						rc.attack(myEnemyTower);
    					}
    					if (rc.canAttack(myEnemyTower)){
    						rc.attack(myEnemyTower);
    					}
    					Direction new_dir = rc.getLocation().directionTo(myEnemyTower).opposite();
    					if (rc.canMove(new_dir)){
    						rc.move(new_dir);
    						rc.setIndicatorString("GOING BACK");
    					}
    					return;
    				}
    			}
    		}
        	if (funcsplayer3.Util.distance(rc.getLocation(), my_target) <= 5 || !rc.isMovementReady()) {
    			Explore.init(rc);
    			Explore.getNewTarget();
        		my_target = Explore.exploreTarget;
    		}
        	rc.setIndicatorString("MOVING TO TARGET AT " + my_target);
        	funcsplayer3.Pathfinding.move(my_target, true);
        	Clock.yield();
        }
    }
    
    public static void getTarget() throws GameActionException {
    	int coord_x = originalLocation.x;
    	int coord_y = originalLocation.y;
    	
    	funcsplayer3.Pathfinding.init(rc);
        funcsplayer3.Pathfinding.initTurn();

    	if (coord_x < coord_y) {
	    	if (coord_x < mapWidth / 2) {
	            Random rand = new Random();
	            int randomInd = rand.nextInt(mapHeight);
	            my_target = new MapLocation(mapWidth - 1, randomInd);
	    	}
	    	else {
	    		Random rand = new Random();
	            int randomInd = rand.nextInt(mapHeight);
	            my_target = new MapLocation(0, randomInd);
	    	}
	        updateEnemyRobots(rc);
    	}
    	else {
    		if (coord_y < mapHeight / 2) {
	            Random rand = new Random();
	            int randomInd = rand.nextInt(mapWidth);
	            my_target = new MapLocation(randomInd, mapHeight - 1);
	    	}
	    	else {
	    		Random rand = new Random();
	    		int randomInd = rand.nextInt(mapWidth);
	            my_target = new MapLocation(randomInd, 0);
	    	}
			// We can also move our code into different methods or classes to better organize it!
	        updateEnemyRobots(rc);
        }
    }

    public static void updateEnemyRobots(RobotController rc) throws GameActionException{
        // Sensing methods can be passed in a radius of -1 to automatically
        // use the largest possible value.
        RobotInfo[] enemyRobots = rc.senseNearbyRobots(-1, rc.getTeam().opponent());
        if (enemyRobots.length != 0){
            rc.setIndicatorString("There are nearby enemy robots! Scary!");
            // Save an array of locations with enemy robots in them for possible future use.
            MapLocation[] enemyLocations = new MapLocation[enemyRobots.length];
            for (int i = 0; i < enemyRobots.length; i++){
                enemyLocations[i] = enemyRobots[i].getLocation();
            }
            RobotInfo[] allyRobots = rc.senseNearbyRobots(-1, rc.getTeam());
            // Occasionally try to tell nearby allies how many enemy robots we see.
            if (rc.getRoundNum() % 20 == 0){
                for (RobotInfo ally : allyRobots){
                    if (rc.canSendMessage(ally.location, enemyRobots.length)){
                        rc.sendMessage(ally.location, enemyRobots.length);
                    }
                }
            }
        }
    }
}

File: python/src/translated/Splasher.py
import random
from battlecode25.stubs import *

class Splasher:

    rc = None
    nearestTower = None
    nearestTowerInfo = None
    prevTarget = None
    my_target = None
    reached_target = False
    hasResource = False

    towerSpotted = False
    myEnemyTower = None

    def __init__(self, rc):
        self.rc = rc

    def play(self):
        if self.towerSpotted:
            if self.rc.can_attack(self.myEnemyTower):
                self.rc.attack(self.myEnemyTower)
            elif self.rc.is_action_ready():
                dir_to_tower = self.rc.get_location().direction_to(self.myEnemyTower)
                if self.rc.can_move(dir_to_tower):
                    self.rc.move(dir_to_tower)
            if self.rc.can_attack(self.myEnemyTower):
                self.rc.attack(self.myEnemyTower)
            if self.rc.sense_robot_at_location(self.myEnemyTower) is None:
                self.towerSpotted = False
            return

        nearby_tiles = self.rc.sense_nearby_map_infos()
        for tile in nearby_tiles:
            potential_tower = self.rc.sense_robot_at_location(tile.get_map_location())
            if tile.has_ruin() and potential_tower is not None:
                if potential_tower.get_team() != self.rc.get_team():
                    # enemy tower
                    self.myEnemyTower = potential_tower.location
                    dir_to_tower = self.rc.get_location().direction_to(self.myEnemyTower)
                    self.towerSpotted = True
                    if self.rc.can_move(dir_to_tower):
                        self.rc.move(dir_to_tower)
                    if self.rc.can_attack(self.myEnemyTower):
                        self.rc.attack(self.myEnemyTower)
                    return

        if Util.distance(self.my_target, self.rc.get_location()) <= 5:
            self.reached_target = True
            new_x = self.map_width - self.my_target.x - 1 if self.original_location.x <= self.map_width / 2 else self.map_width - self.my_target.x + 1
            new_y = self.map_height - self.my_target.y - 1 if self.original_location.y <= self.map_height / 2 else self.map_height - self.my_target.y + 1
            self.my_target = MapLocation(new_x, new_y)
        elif not self.reached_target:
            self.get_target()

        self.rc.set_indicator_string(f"MOVING TO TARGET AT {self.my_target}")
        Pathfinding.move(self.my_target, True)

    def get_target(self):
        coord_x = self.original_location.x
        coord_y = self.original_location.y

        Pathfinding.init(self.rc)
        Pathfinding.init_turn()

        if coord_x < coord_y:
            if coord_x < self.map_width / 2:
                random_ind = random.randint(0, self.map_height - 1)
                self.my_target = MapLocation(self.map_width - 1, random_ind)
            else:
                random_ind = random.randint(0, self.map_height - 1)
                self.my_target = MapLocation(0, random_ind)
            self.update_enemy_robots(self.rc)
        else:
            if coord_y < self.map_height / 2:
                random_ind = random.randint(0, self.map_width - 1)
                self.my_target = MapLocation(random_ind, self.map_height - 1)
            else:
                random_ind = random.randint(0, self.map_width - 1)
                self.my_target = MapLocation(random_ind, 0)
            self.update_enemy_robots(self.rc)

    def update_enemy_robots(self, rc):
        enemy_robots = rc.sense_nearby_robots(-1, rc.get_team().opponent())
        if enemy_robots:
            self.rc.set_indicator_string("There are nearby enemy robots! Scary!")
            enemy_locations = [robot.get_location() for robot in enemy_robots]

            ally_robots = rc.sense_nearby_robots(-1, rc.get_team())
            if rc.get_round_num() % 20 == 0:
                for ally in ally_robots:
                    if rc.can_send_message(ally.location, len(enemy_robots)):
                        rc.send_message(ally.location, len(enemy_robots))
